### ISSUE TITLE
del Tensor.grad segfaults

### ISSUE THREAD

**Issue by colesbury** [pytorch/pytorch#16471]
## 🐛 Bug
Attempting to delete the `grad` attribute of a Tensor from Python results in a segmentation fault.

## To Reproduce
```python
import torch
x = torch.randn(3, 3, requires_grad=True)
del x.grad
```

Python 3, PyTorch master (Jan 25, 2019)

## Additional context
The delete gets translated by the Python C API to a set attribute with `nullptr` (as opposed to `Py_None`) here:

https://github.com/pytorch/pytorch/blob/e3c0926c449dd8645583216c96a981d43efe2aef/torch/csrc/autograd/python_variable.cpp#L234-L244

The `THPVariable_Check` on a nullptr segfaults.

**Comment by vishwakftw** [comment]
Is resetting the `grad` attribute (as done in the case of `PyNone`) the expected behavior here?

**Comment by bhushan23** [comment]
> Is resetting the `grad` attribute (as done in the case of `PyNone`) the expected behavior here?

Yes. py_grad should be Py_None and not nullptr

**Comment by bhushan23** [comment]
I am not table to find the start point for del
Where does delete gets translated into C API?

Following is the stack trace
```
#0  0x00007ffff7e61710 in PyObject_IsInstance () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0                                  
#1  0x00007fffe85e0402 in THPVariable_Check(_object*) () from /home/bhushan/miniconda3/envs/pytorch-dev/lib/python2.7/site-packages/torch/lib/libtorch_python.so
#2  0x00007fffe8760f29 in THPVariable_set_grad(THPVariable*, _object*) () from /home/bhushan/miniconda3/envs/pytorch-dev/lib/python2.7/site-packages/torch/lib/libtorch_python.so
#3  0x00007ffff7e75e11 in getset_set () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#4  0x00007ffff7e9d41b in _PyObject_GenericSetAttrWithDict () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#5  0x00007ffff7e9cf30 in PyObject_SetAttr () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#6  0x00007ffff7ef6ace in PyEval_EvalFrameEx () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#7  0x00007ffff7efc6c9 in PyEval_EvalCodeEx () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#8  0x00007ffff7efc8ea in PyEval_EvalCode () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#9  0x00007ffff7f15bad in run_mod () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#10 0x00007ffff7f16d28 in PyRun_FileExFlags () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#11 0x00007ffff7f17f48 in PyRun_SimpleFileExFlags () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#12 0x00007ffff7f2a16c in Py_Main () from /home/bhushan/miniconda3/envs/pytorch-dev/bin/../lib/libpython2.7.so.1.0
#13 0x00007ffff7041b97 in __libc_start_main (main=0x555555555050 <main>, argc=2, argv=0x7fffffffdd98, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffdd88)
    at ../csu/libc-start.c:310
#14 0x000055555555507f in _start () at ../sysdeps/x86_64/elf/start.S:103
```

**Comment by bhushan23** [comment]
#16525 is ready to merge

**Comment by vishwakftw** [comment]
This can be closed now, since the fix has landed.

**Comment by soumith** [comment]
closed via #16525
